File: Algorithms-April-2016/[EXAMS]/Algorithms-Sample-Exam/02. Bridges/Bridges.cs
namespace Bridges
{
    using System;
    using System.Collections.Generic;
    using System.Linq;

    public class Bridges
    {
        private static int[,] bridgesCount;

        public static void Main()
        {
            int[] north = Console.ReadLine().Split().Select(int.Parse).ToArray();
            int[] south = Console.ReadLine().Split().Select(int.Parse).ToArray();

            bridgesCount = new int[north.Length + 1, south.Length + 1];
            for (int northIndex = 1; northIndex < bridgesCount.GetLength(0); northIndex++)
            {
                for (int southIndex = 1; southIndex < bridgesCount.GetLength(1); southIndex++)
                {
                    if (north[northIndex - 1] == south[southIndex - 1])
                    {
                        bridgesCount[northIndex, southIndex] = GetMax(
                            bridgesCount[northIndex - 1, southIndex],
                            bridgesCount[northIndex, southIndex - 1],
                            bridgesCount[northIndex - 1, southIndex - 1]);
                        bridgesCount[northIndex, southIndex]++;
                    }
                    else
                    {
                        bridgesCount[northIndex, southIndex] = GetMax(
                            bridgesCount[northIndex - 1, southIndex],
                            bridgesCount[northIndex, southIndex - 1],
                            bridgesCount[northIndex - 1, southIndex - 1]);
                    }
                }
            }

            Console.WriteLine(bridgesCount[north.Length, south.Length]);
        }

        private static int GetMax(int firstNum, int secondNum, int thirdNum)
        {
            int currentMax = Math.Max(firstNum, secondNum);
            return Math.Max(currentMax, thirdNum);
        }
    }
}

File: Algorithms-April-2016/[EXAMS]/Algorithms-Sample-Exam/Bridges.java
import java.util.Scanner;

public class Bridges {

    private static int[][] matrix;

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        String[] northInput = scanner.nextLine().split(" ");
        int[] north = new int[northInput.length];
        for (int i = 0; i < northInput.length; i++) {
            north[i] = Integer.valueOf(northInput[i]);
        }

        String[] southInput = scanner.nextLine().split(" ");
        int[] south = new int[southInput.length];
        for (int i = 0; i < southInput.length; i++) {
            south[i] = Integer.valueOf(southInput[i]);
        }

        matrix = new int[north.length + 1][south.length + 1];
        for (int i = 1; i < matrix.length; i++) {
            for (int j = 1; j < matrix[i].length; j++) {
                if (north[i - 1] == south[j - 1]) {
                    matrix[i][j] = getMax(matrix[i - 1][j], matrix[i][j - 1], matrix[i - 1][j - 1]) + 1;
                } else {
                    matrix[i][j] = getMax(matrix[i - 1][j], matrix[i][j - 1], matrix[i - 1][j - 1]);
                }
            }
        }

        System.out.println((matrix[north.length][south.length]));
    }

    private static int getMax(int p1, int p2, int p3) {
        int max = Math.max(p1, p2);
        return Math.max(max, p3);
    }
}